dm/docs/RFCS/20211130_enhanced_pre_checker.md

```diff
@@ -0,0 +1,86 @@
+* For schema_of_shard_tables, whether it's a pessimistic task or an optimistic task, we all simply check consistency of schema by comparing all sharding tables’ structures. For optimistic mode, we can do better.
+* Version checker checks that MySQLVersion >= 5.6.0 and MariadbVersion >= 10.1.2 before. However, we find more and more incompatibility problems in MySQLVersion >= 8.0.0 and Mariadb. In view of supporting MySQL 8.0 and Mariadb is experimental yet, checker will report a warning for them.
+
+## Proposal
+
+### Speed ​​up check
+
+1. Support concurrent check
+    - table_schema
+    - schema_of_shard_tables
+    - auto_increment_ID
+2. We can adjust the concurrency by table numbers.
+
+### Optimize some check
+
+1. Auto_increment_ID only be checked in sharding mode.
+    - If table exists auto increment ID, report a warning to user and tell them the method that can resolve the PK/UK conflict;
+        1. If you set PK to AUTO INCREMENT, you must make sure that the primary key in sharding tables is not duplicated;
+        2. If sharding tables have duplicated PK, you can refer to [document](https://docs.pingcap.com/tidb-data-migration/stable/shard-merge-best-practices#handle-conflicts-of-auto-increment-primary-key).
+    - And if they have finished the resolving, such as manually created the table and removed AUTO INCREMENT, we should not report the warning.
+2. Dump_privilege will check different privileges according to different [consistency](https://docs.pingcap.com/tidb/stable/dumpling-overview#adjust-dumplings-data-consistency-options) on source.
+    - For all consistency, we will check
+        - SELECT (only INFORMATION_SCHEMA's tables and dump tables)
+    - For flush consistency:
+        - RELOAD (global)
+    - For flush/lock consistency:
+3. Add OnlineDDLChecker to check if a DDL of tables in allow list exists in online-ddl stage when DM task is all mode and online-ddl is true.
+4. Enhance schema_of_shard_tables. 
+    - If a task has passed the pre-checking when starting and exited, DM should keep the consistency during the task running. So we **don't check it** when restart the task.
+    - If there does not exist checkpoints:
+        - For all/full mode (pessimistic task): we keep **the original check**;
+        - For all/full mode (optimistic task): we check whether the shard tables schema meets **the definition of [Optimistic Schema Compatibility](20191209_optimistic_ddl.md)**. If that meets, we can create tables by the compatible schema in the dump stage.
+        - For incremental mode: **not check** the sharding tables’ schema, because the table schema obtained from show create table is not the schema at the point of binlog.
+5. Version checker will report a warning in the following cases:
+    - MySQL < 5.6.0
+    - MySQL >= 8.0.0
+    - Mariadb
+    - Others
```

Comment:
"Others" will be hard to check because they will pretend themself as MySQL and we need case-by-case method to check them.